### PR TITLE
prevent block updates when pending

### LINKED ISSUE #13283: prevent further updates to an existing Block when it's guid is present in a BlockSubmission
follow on from  #13201 

We should show/block further edits in existing Block page when a BlockSubmission is outstanding.  Ideally the same restriction should be shown in the GUID input view too if a matching GUID is in an outstanding BlockSubmission

### PR DESCRIPTION
**Pull request by eviljeff**
fixes #13283

### CODE REVIEW

File: src/olympia/blocklist/templates/blocklist/block_change_form.html [hunk 1]
```diff
@@ -4,13 +4,5 @@
 {% block submit_buttons_bottom %}
   {# This is so we can submit directly to the blocksubmission add view - it expect an input_guids field #}
   <input type="hidden" name="input_guids" value="{{ original.guid }}">
-  <div class="submit-row">
```
Comment (bobsilverberg) [inline src/olympia/blocklist/templates/blocklist/block_change_form.html]: I don't understand this change. I can see that code that is similar to (but not exactly like) this exists in `src/olympia/blocklist/templates/blocklist/block_submission_change_form.html`, but why is this being removed from here. Also, what does `{{ block.super }}` do? Where is `"admin/change_form.html"` that this extends?

File: src/olympia/blocklist/tests/test_admin.py [hunk 2]
```diff
@@ -1196,6 +1208,7 @@ def test_can_not_edit_without_permission(self):
         # Try to edit the block anyway
         response = self.client.post(
             self.change_url, {
+                'input_guids': self.block.guid,
```
Comment (bobsilverberg) [inline src/olympia/blocklist/tests/test_admin.py]: Why was this change necessary in this patch?

File: src/olympia/blocklist/admin.py [hunk 3]
```diff
@@ -502,9 +502,19 @@ def url_link(self, obj):
         return format_html('<a href="{}">{}</a>', obj.url, obj.url)
 
     def block_history(self, obj):
+        submission = BlockSubmission.get_submission_from_guid(obj.guid).first()
```
Comment (bobsilverberg) [inline src/olympia/blocklist/admin.py]: I assume from the `first()` that it's it possible that `get_submission_from_guid` could return more than one submission. Is it sufficient to only output the first pending submission found?

File: src/olympia/blocklist/admin.py [hunk 4]
```diff
@@ -502,9 +502,19 @@ def url_link(self, obj):
...
+                date=submission.modified.date(),
+                url=reverse(
+                    'admin:blocklist_blocksubmission_change',
+                    args=(submission.id,)))
+            if submission else '')
         return format_block_history(
             ActivityLog.objects.for_guidblock(obj.guid).filter(
-                action__in=Block.ACTIVITY_IDS).order_by('created'))
+                action__in=Block.ACTIVITY_IDS).order_by('created'),
+            additional_logs=submission_log)
```
Comment (bobsilverberg) [inline src/olympia/blocklist/admin.py]: Should there be a test for this new code? I don't see one anywhere.

File: src/olympia/blocklist/admin.py [hunk 5]
```diff
@@ -534,6 +544,12 @@ def get_fieldsets(self, request, obj):
 
         return (details, history, edit)
 
+    def has_change_permission(self, request, obj=None):
+        if obj and BlockSubmission.get_submission_from_guid(obj.guid):
```
Comment (bobsilverberg) [inline src/olympia/blocklist/admin.py]: Should there be a test for this, or is it considered covered by `test_cannot_edit_when_guid_in_blocksubmission`?

File: src/olympia/blocklist/admin.py [hunk 6]
```diff
@@ -550,10 +566,11 @@ def delete_queryset(self, request, queryset):
 
     def get_form(self, request, obj=None, change=False, **kwargs):
         form = super().get_form(request, obj=obj, change=change, **kwargs)
-        form.base_fields['min_version'].choices = _get_version_choices(
-            obj, 'min_version')
-        form.base_fields['max_version'].choices = _get_version_choices(
-            obj, 'max_version')
+        if 'min_version' in form.base_fields:
```
Comment (bobsilverberg) [inline src/olympia/blocklist/admin.py]: Why is this change necessary for this patch? I don't see any changes that would cause `form.base_fields` to not have `min_version`.

File: src/olympia/blocklist/admin.py [hunk 7]
```diff
@@ -573,6 +590,8 @@ def changeform_view(self, request, obj_id=None, form_url='',
                 return HttpResponseTemporaryRedirect(
                     reverse('admin:blocklist_blocksubmission_add'))
 
+        extra_context = extra_context or {}
```
Comment (bobsilverberg) [inline src/olympia/blocklist/admin.py]: I'm not really sure what the `extra_context` does or where it comes from, but instead of this could you just default `extra_context` to `{}` instead of `None` on line 582?

File: src/olympia/blocklist/forms.py [hunk 8]
```diff
@@ -39,10 +39,18 @@ class MultiAddForm(MultiGUIDInputForm):
...
             if not blk and not Addon.unfiltered.filter(guid=guid).exists():
-                raise ValidationError(
+                errors.append(ValidationError(
                     _('Addon with GUID %(guid)s does not exist'),
-                    params={'guid': guid})
+                    params={'guid': guid}))
+        for guid in guids:
+            if BlockSubmission.get_submission_from_guid(guid):
+                errors.append(ValidationError(
+                    _('GUID %(guid)s is already in pending BlockSubmission'),
```
Comment (bobsilverberg) [inline src/olympia/blocklist/forms.py]: Should the message read: "GUID %(guid)s is already in *a* pending BlockSubmission"?

File: src/olympia/blocklist/tests/test_models.py [hunk 9]
```diff
@@ -64,3 +64,25 @@ def test_is_save_to_blocks_permitted(self):
         # Except when that's not enforced locally
         with override_settings(DEBUG=True):
             assert block.is_save_to_blocks_permitted
+
+    def test_get_submission_from_guid(self):
```
Comment (bobsilverberg) [inline src/olympia/blocklist/tests/test_models.py]: For safety's sake, is it worth also having a test that proves that a `guid` that does not exist in a submission returns an empty list?

File: src/olympia/blocklist/models.py [hunk 10]
```diff
@@ -383,6 +387,12 @@ def save_to_blocks(self):
 
         self.update(signoff_state=self.SIGNOFF_PUBLISHED)
 
+    @classmethod
+    def get_submission_from_guid(cls, guid, states=SIGNOFF_STATES_FINISHED):
+        return (
```
Comment (bobsilverberg) [inline src/olympia/blocklist/models.py]: Will this return an empty list if nothing is found, or `None`?

File: src/olympia/blocklist/admin.py (hunk 5 above)
Comment (bobsilverberg) [inline src/olympia/blocklist/admin.py]: Also, I asked this elsewhere, but I assume that `get_submission_from_guid` returns `None` if none are found and not an empty list?

File: src/olympia/blocklist/admin.py (hunk 3 above)
Comment (eviljeff) [reply src/olympia/blocklist/admin.py]: yeah, it returns a queryset.  A guid should only be in one BlockSubmission at once (because adding it would otherwise have been blocked by the changes in this patch).  `.first()` is just an easy (lazy) shortcut because it will either return one `BlockSubmission` instance or `None`.

File: src/olympia/blocklist/admin.py (hunk 4 above)
Comment (eviljeff) [reply src/olympia/blocklist/admin.py]: https://github.com/mozilla/addons-server/pull/13479/files#diff-55adb0dc3ea4f62a69a70ccd80833668R1238-R1240

File: src/olympia/blocklist/admin.py (hunk 5 above)
Comment (eviljeff) [reply src/olympia/blocklist/admin.py]: that covers it - when `has_change_permission` is `False` the page is read only.

Comment (eviljeff) [reply src/olympia/blocklist/admin.py]: it returns a queryset which is falsey when there are no results.  (an empty list is also falsey in python)

File: src/olympia/blocklist/admin.py (hunk 6 above)
Comment (eviljeff) [reply src/olympia/blocklist/admin.py]: when `has_change_permission` is `False` the form is empty so `.base_fields` doesn't contain the min/max fields.

File: src/olympia/blocklist/admin.py (hunk 7 above)
Comment (eviljeff) [reply src/olympia/blocklist/admin.py]: > I'm not really sure what the extra_context does or where it comes from, 

It's extra data that will be passed to the templates when rendering (the "context" is the dict containing all the dynamic content needed to render).

> but instead of this could you just default extra_context to {} instead of None on line 582?  

In python the default args are reused for the function so if you define a mutable object and modify it then it will still be there _next_ time that function is called.  Currently in this function we're always modifying the object in the same way and only referencing it afterwards but imo it's good practise to take that common gotcha into account (if any of the code above was changed to reference `extra_context` it'd lead to a confusing bug).

File: src/olympia/blocklist/models.py (hunk 10 above)
Comment (eviljeff) [reply src/olympia/blocklist/models.py]: it returns a queryset which is falsey when there are no results.

File: src/olympia/blocklist/models.py [hunk 11]
```diff
@@ -383,6 +387,12 @@ def save_to_blocks(self):
...
+            cls.objects.exclude(signoff_state__in=states)
+                       .filter(to_block__contains=f'"{guid}"'))
```
[the remark this answers is not in the document]
Comment (eviljeff) [reply src/olympia/blocklist/models.py]: the field is storing JSON data so the `__contains` is just a way of matching it - there isn't native support for JSON in the database.  The `"..."` should ensure that only full guids are matched.  (Now I see there's an edge case if a guid is entered in an invalid format, but it'd be an error thrown on invalid input so no real harm there)

File: src/olympia/blocklist/templates/blocklist/block_change_form.html (hunk 1 above)
Comment (eviljeff) [reply src/olympia/blocklist/templates/blocklist/block_change_form.html]: `{{ block.super }}` means use the parent definition of `{% block submit_buttons_bottom %}` (like `super()` inside a python object function).  `admin/change_form.html` is the original django admin template https://github.com/django/django/blob/2.2.10/django/contrib/admin/templates/admin/change_form.html

I removed my custom template fragment because I realised it wasn't needed any longer.

File: src/olympia/blocklist/tests/test_admin.py (hunk 2 above)
Comment (eviljeff) [reply src/olympia/blocklist/tests/test_admin.py]: It wasn't strictly necessary but I noticed it so drive-by fixed it.  The `.post` would still fail anyway but if the permission checks were broken in subsequent patch then, without `input_guids`, it would fail in a different way (it'd 200 back to the same page because the form data was broken) which could be confusing.

File: src/olympia/blocklist/tests/test_models.py (hunk 9 above)
Comment (eviljeff) [reply src/olympia/blocklist/tests/test_models.py]: https://github.com/mozilla/addons-server/pull/13479/commits/7595cf2bf80cbc3035ebc93e7ecbe3b1947e81c3

File: src/olympia/blocklist/forms.py (hunk 8 above)
Comment (eviljeff) [reply src/olympia/blocklist/forms.py]: https://github.com/mozilla/addons-server/pull/13479/commits/7595cf2bf80cbc3035ebc93e7ecbe3b1947e81c3
